test/integration/network_extension_discovery_integration_test.cc

```diff
@@ -0,0 +1,707 @@
+  // The two filters drain 4 + 5  bytes.
+  sendDataVerifyResults(9);
+}
+
+// Testing it works with two static network filter configuration.
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, TwoStaticFilters) {
+  addFilterChain();
+  addStaticFilter("foo", 3);
+  addStaticFilter("bar", 5);
+  initialize();
+
+  // filter drain 3+5 bytes.
+  sendDataVerifyResults(8);
+}
+
+// Testing it works with mixed static/dynamic network filter configuration.
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, TwoDynamicTwoStaticFilterMixed) {
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter(filter_name_, false);
+  addStaticFilter("bar", 2);
+  addDynamicFilter(filter_name_, true);
+  addStaticFilter("foobar", 2);
+  initialize();
+
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  sendXdsResponse(filter_name_, "1", 3);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_reload", 1);
+  // filter drain 3 + 2 + 3 + 2 bytes.
+  sendDataVerifyResults(10);
+}
+
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, DynamicStaticFilterMixedDifferentOrder) {
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addStaticFilter("bar", 2);
+  addStaticFilter("baz", 2);
+  addDynamicFilter(filter_name_, true);
+  addDynamicFilter(filter_name_, false);
+  initialize();
+
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  sendXdsResponse(filter_name_, "1", 2);
+  test_server_->waitForCounterGe(
+      "extension_config_discovery.network_filter." + filter_name_ + ".config_reload", 1);
+  // filter drain 2 + 2 + 2 + 2 bytes.
+  sendDataVerifyResults(8);
+}
+
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, DestroyDuringInit) {
+  // If rate limiting is enabled on the config source, gRPC mux drainage updates the requests
+  // queue size on destruction. The update calls out to stats scope nested under the extension
+  // config subscription stats scope. This test verifies that the stats scope outlasts the gRPC
+  // subscription.
+  on_server_init_function_ = [&]() { waitXdsStream(); };
+  addFilterChain();
+  addDynamicFilter(filter_name_, false, true);
+  initialize();
+
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+
+  test_server_.reset();
+  auto result = ecds_connection_->waitForDisconnect();
+  ASSERT_TRUE(result);
+  ecds_connection_.reset();
+}
+
+// Validate that a listener update should fail if the subscribed extension configuration make filter
+// terminal but the filter position is not at the last position at filter chain.
+// There would be total of 2 filters in the chain: 'foo' and 'tcp_proxy' and both are marked
+// terminal.
+TEST_P(NetworkExtensionDiscoveryIntegrationTest, BasicFailTerminalFilterNotAtEndOfFilterChain) {
```

Comment:
FWIW, I don't think that listener and network filters should be terminal, but I guess this is "inherited" from http filters.

Reply:
``tcp_proxy``/``hcm`` for example are defined as terminal, i.e can only be used as last network filter

Reply:
Ack.
I guess the use of the term [is_terminal here](https://github.com/envoyproxy/envoy/pull/28477/files#diff-791ac42e27b3c608b85afdc77097fe7a1e99b6e590be8fbbce60f5f165c91c4fR88) is somewhat misleading.